storage/src/store/mod.rs

```diff
@@ -0,0 +1,919 @@
+                        // before they are written to the snapshot.
+                        uncommitted_updates = uncommitted_updates
+                            .into_iter()
+                            .filter(|(_, _, _, k)| *k != key)
+                            .collect::<Vec<_>>();
+
+                        let Some(old_loc) = self.get_loc(&key).await? else {
+                            continue;
+                        };
+
+                        Self::delete_loc(&mut self.snapshot, &key, old_loc);
+                    }
+                    Operation::Commit(floor_loc) => {
+                        // Bump the inactivity floor
+                        self.inactivity_floor_loc = floor_loc;
+                        last_commit_loc = Some((section, offset));
+
+                        // Flush all uncommitted update operations to the snapshot
+                        for (pos, _, _, key) in uncommitted_updates.iter() {
+                            let Some(old_loc) = self.get_loc(key).await? else {
+                                self.snapshot.insert(key, *pos);
+                                continue;
+                            };
+
+                            Self::update_loc(&mut self.snapshot, key, old_loc, *pos);
+                        }
+
+                        // Clear uncommitted operations
+                        uncommitted_updates.clear();
+                        uncommitted_ops = 0;
+                    }
+                }
+            }
+        }
+
+        if uncommitted_ops > 0 {
+            if let Some((last_section, last_offset)) = last_commit_loc {
+                let last_uncommitted_loc = self.op_count - uncommitted_ops;
+                self.op_count = last_uncommitted_loc;
+
+                try_join!(
+                    self.log.rewind(last_section, last_offset as u64 + 1),
+                    self.locations.rewind(last_uncommitted_loc)
+                )?;
+            } else {
+                // No commits found; Drain log entirely.
+                self.op_count = 0;
+                try_join!(self.log.rewind(0, 0), self.locations.rewind(0))?;
+            }
+        }
+
+        self.oldest_retained_loc = oldest_retained_loc.unwrap_or(0);
+
+        assert_eq!(self.op_count, self.locations.size().await?);
+        assert_eq!(self.uncommitted_ops, 0);
+
+        Ok(self)
+    }
+
+    /// Append the operation to the log. The `commit` method must be called to make any applied operation
+    /// persistent & recoverable.
+    async fn apply_op(&mut self, op: Operation<K, V>) -> Result<u32, Error> {
+        let current_section = self.op_count / self.log_items_per_section;
+
+        self.uncommitted_ops += 1;
+
+        // Append the operation to the entry log, and the offset to the locations log.
+        let (offset, _size) = self.log.append(current_section, op).await?;
+        self.locations.append(offset.into()).await?;
```

Comment:
This is a clever optimization to only store offset here. I think it deserves a little additional documentation as to why we do this (using the "op number" to derive the section later).

Reply:
Added